### ISSUE TITLE
Query with Cosine similarity

### ISSUE THREAD

**Issue by hegdenagaraj4** [iit-cs579/main#326]
I have a doubt. If I understand correctly, weighted average is obtained after we have the similarity. And for the cosine similarity I will have to pass the userprofile and the item profile. I would have the matrix from the item profile after calculating the tf-idf for each movie. But I am not able to understand how to build the user profile in order to get the cosine similarity. Please explain.

**Comment by vmanikes** [comment]
Same problem here! I am not able to understand how to build the user profile

**Comment by hegdenagaraj4** [comment]
@ClaytonTurner  @AndrewLu1992  @aronwc  If I understand correctly, for each user, in the movie row, I will be calculating the cosine similarity. For the rows with positive cosine similarity I will calculate the weighted average. Now this calculated weight is compared to the rating in the ratings_test. Please correct me if I am wrong. 
The tf-idf is the document profile. So does it mean for a movie, for each user for that row, call cosine similarity(user_ratings (user_vector), tf-idf_matrix(movie_row_vector))?

**Comment by joezhang2** [comment]
Lets say in the training data, user A has the reviewed these movies: 
(Finding Dory: 8),
(Star Wars 7: 7), 
(Remember the Titans: 6),
(Lincoln: 10) 
and we want to predict their rating for Finding Nemo in the test data. 

We want to first compute the cosine similarity between Finding Nemo and the 3 movies movies reviewed in the training data. So for example:
(Finding Nemo, Finding Dory: .8),
(Finding Nemo, Star Wars 7: .3),
(Finding Nemo, Remember the Titans: .1),
(Finding Nemo, Lincoln: 0). 

Now we can predict the rating using a weighted average from the cosine similarity and the users ratings: 
(.8 * 8 +.3 * 7 + .1 * 6 + 0 * 10)/ (.8 + .3 + .1 + 0)

Note: If we get a 0 for every movie from the cosine similarity, just average the ratings as is

**Comment by hegdenagaraj4** [comment]
@joezhang2 okay so we are matching the user column with the movie row correct? I will try that. How will  Cosine similarity do the normalization?When I pass two vectors does it take the numerical values available inside the vector and calculate? Because movies profile has tf-idf matrix only as numerical data, and ratings_train will have user ratings.  I am trying to understand the importance of tf-idf that we calculated.

**Comment by joezhang2** [comment]
In featurize, we use the tf-idf to make a 1xN matrix, where N is the length of the vocabulary. Once we have a matrix for each movie, we can compute the cosine similarity between two movies using the matrix for each movie. This serves as a sort of measure for how similar two movies are. 

So if two movies are similar, their ratings should be similar as well, e.g. if I like cartoons and hate horror movies, my predicted rating for a cartoon movie should be high and my predicted rating for a horror movie should be low.

We already have a user's ratings for other movies from the training data, so we just need to select the ratings for movies similar to the one we want to predict, which we accomplish using the cosine similarity. I can come up with a more thorough example if you are still confused

**Comment by hegdenagaraj4** [comment]
@joezhang2 awesome. I think I kind of understood. I will try. Thanks :)

**Comment by vmanikes** [comment]
@joezhang2 That was a nice explanation but I am confused with this sentence, could you please elaborate it 

> We already have a user's ratings for other movies from the training data, so we just need to select the ratings for movies similar to the one we want to predict, which we accomplish using the cosine similarity

**Comment by ddabhane** [comment]
@vmanikes It means only those movie ratings from training data is selected for weighted average whose cosine similarity with the movie (test data) is positive.

**Comment by AndrewLu1992** [comment]
@joezhang2 You explained very well. To predict the user rating of movie, use cosine similarity as weight and character the movie as a vector of similar other movies.

**Comment by joezhang2** [comment]
@ddabhane Yup! 

@vmanikes, I was just elaborating on the logic behind the math in the assignment. Here is a more complete example:
Say someone rated:
> Lord of the Rings: (fantasy, live-action), rating: 10
Frozen: (fantasy, cartoon, children), rating: 6 
Lincoln: (biography), rating: 2 

and we are asked to predict how they rate Chronicles of Narnia (fantasy, live-action, children).

We need to create a vocab: (biography, cartoon, children, fantasy, live-action) and featurize the vectors. The numbers correspond to the vocab.
> Lord of the Rings (0, 0, 0, 1, 1)
Frozen (0, 1, 1, 0, 0)
Lincoln (1, 0, 0, 0, 0)
Chronicles of Narnia (0, 0, 1, 1, 1)

Now we want to predict a score for Chronicles of Narnia. We first find the cosine similarity score between the 3 movies that the user rated and Chronicles of Narnia. I am making up numbers and not doing the actual math here.
> Lord of the Rings: .8
Frozen: .7
Lincoln: 0

Now we have a score of how similar movies are so we can weigh the ratings for the movies by their cosine similarity scores:
> Lord of the Rings: .8 * 10
Frozen: .7 * 6
Lincoln: 0 * 2

Add up the scores:
>8 + 4.2 + 0 = 12.2

Now divide by the sum of the weights to get a weighted average prediction:
>12.2/(.8 + .7 + 0) = approximately 8

I hope the logic behind what is going on is clear as that is how we can understand the limitations and the appropriate uses for method.

**Comment by vmanikes** [comment]
Thank you @joezhang2 @ddabhane

**Comment by ddabhane** [comment]
For calculating dot(a,b) in cosine_sim, one of the input vector needs to be transpose. So either dot(a.transpose,b) or dot(a,b.transpose).
So which vector do we need to take transpose?

**Comment by ClaytonTurner** [comment]
It shouldn't matter which you transpose for a dot product

**Comment by ddabhane** [comment]
@ClaytonTurner The value of dot(a.transpose,b) and  dot(a,b.transpose) are different, so what the function returns will also be different.

**Comment by ClaytonTurner** [comment]
That's a bit confusing as we are given 2 vectors for this (a csr matrix of size 1,n is really just a vector).

if our matrices were (gonna just write them as lists for simplicity's sake): 
a = [2,4,6]
b = [-1,0,1]

then the dot product of a and b equals: 2*-1 + 4*0 + 6*1 
and this is equivalent to b and a: -1*2 + 0*4 + 1*6

**Comment by ddabhane** [comment]
@ClaytonTurner Thanks, you are right. I converted a and b into array and it worked.

**Comment by thirumalrao** [comment]
In some instances, cosine_sim may result in [nan]. Is this fine or should this be considered as 0?
 
An example : if a = [0.3,0.3,0] and b = [0,0,0], then dot product will be 0, and L2 norm of b will also be 0. So, cosine_sim would be 0/0 = nan

**Comment by joezhang2** [comment]
@thirumalrao The only way to get a zero vector like in your example requires a movie to have an empty string in the genres column. I just checked the movies and did not find any movies like that. 

When I calculated the weighted averages, one of my test values returned 'nan'. I resolved this be manually coding the dot product/l2norms for the cosine_sim function instead of calling on a library and my problem went away.

**Comment by thirumalrao** [comment]
@joezhang2 - Lets take an example - ([123, 'Horror|Romance|Sci-fi'], [456, 'Sci-Fi|Sci-fi'])
where a = ['Horror|Romance|Sci-fi']
b = ['Sci-Fi|Sci-fi']
in this case, what would be the tf-idf for b ?
tfidf = term_counts[i]/max-k * math.log10(N/df(i))

term_counts['Sci-fi'] = 2 
max-k = 2
N = 2
df('Sci-fi') = 2

so, tf-idf = 2/2 * log10(1) = 0
Given that there is only 'Sci-fi' in b, featurize will convert this into a csr matrix [0, 0, 0] -- isn't it ?

Now csr-matrix for b is [0 0 0]
If there is another matrix say [0 1 1] then dot product these 2 vectors will be 0.
L2Norm of b is also 0.
Wouldn't this result in cosine_sim value to be 0/0 => nan ?

**Comment by Tsreenivasan** [comment]
@thirumalrao 

td-idf is a value computed for 'b' against all the data in the example set. [ measure of how important a word is to a document in the collection]
realistically, if every movie in the review system was of Sci-fi genre,  then using genre as token is highly biased for our prediction, hence  the tf-idf algorithm eliminates such scenarios of similarities, by treating 'Sci-fi' as not important to the document.

the only way the example you stated will yield a zero tf-idf value is, term in question be present in every document of the training set.

Apologies if the question was only related to the nan result! my assumption was that the question was regarding the tf-idf numerical.

**Comment by thirumalrao** [comment]
@Tsreenivasan - you answered my query indirectly. Yes, in my example, term in question is there in both documents. Probably this won't occur in the real dataset.
